core/src/test/java/com/github/mizool/core/configuration/TestConfig.java

```diff
@@ -0,0 +1,336 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.github.mizool.core.configuration;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.catchThrowable;
+
+import java.math.BigDecimal;
+import java.net.URL;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Paths;
+import java.time.Duration;
+import java.time.OffsetDateTime;
+import java.time.Period;
+import java.time.ZoneId;
+import java.time.ZoneOffset;
+import java.util.List;
+import java.util.Optional;
+import java.util.Properties;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+import lombok.RequiredArgsConstructor;
+import lombok.ToString;
+
+import org.testng.annotations.BeforeMethod;
+import org.testng.annotations.DataProvider;
+import org.testng.annotations.Test;
+
+import com.github.mizool.core.exception.ConfigurationException;
+import com.google.common.collect.ImmutableList;
+
+public class TestConfig
+{
+    private static final String VALUE = "foo";
+
+    private Properties testProperties;
+
+    @BeforeMethod
+    void setUp()
+    {
+        testProperties = new Properties();
+        testProperties.setProperty("answer", "forty-two");
+        testProperties.setProperty("something", "interesting");
+
+        testProperties.setProperty("a", VALUE);
+        testProperties.setProperty("a.b", VALUE);
+        testProperties.setProperty("a.b.c", VALUE);
+    }
+
+    @Test
+    public void testProperties()
+    {
+        RootNode root = Config.from(testProperties);
+
+        assertThat(root.child("answer").stringValue().obtain()).isEqualTo("forty-two");
+        assertThat(root.child("something").stringValue().obtain()).isEqualTo("interesting");
+    }
+
+    @Test
+    public void testPropertiesWithDefaults()
+    {
+        Properties propertiesWithDefaults = new Properties(testProperties);
+        propertiesWithDefaults.setProperty("greeting", "Hello, World!");
+
+        RootNode root = Config.from(propertiesWithDefaults);
+        assertThat(root.child("answer").stringValue().obtain()).isEqualTo("forty-two");
+        assertThat(root.child("something").stringValue().obtain()).isEqualTo("interesting");
+        assertThat(root.child("greeting").stringValue().obtain()).isEqualTo("Hello, World!");
+    }
+
+    @Test
+    public void testSystemProperties()
+    {
+        RootNode root = Config.systemProperties();
+        String lineSeparator = root.child("line.separator").stringValue().obtain();
+        assertThat(lineSeparator).isNotEmpty();
```

Comment:
Since `obtain` throws an exception, change that to `read` and base the assertion on the presence of the value.

Also note that the `read` code already filters empty strings. That might be something you want to test separately.

Reply:
> Also note that the read code already filters empty strings. That might be something you want to test separately.

This was in fact already covered by the `TestValue.testValues()` case called "empty". However, your comment made me realize that it was not obvious enough to readers, so I refactored it into its own test method.